beginner_source/aws_distributed_training_tutorial.py

```diff
@@ -0,0 +1,675 @@
+# handles the averaging of gradients across the world, so we do not have
+# to explicitly average the gradients in the training step.
+# 
+# It is important to note that this is a blocking function, meaning
+# program execution will wait at this function until *world\_size*
+# processes have joined the process group. Also, notice we pass our device
+# ids list as a parameter which contains the local rank (i.e. GPU) we are
+# using. Finally, we specify the loss function and optimizer to train the
+# model with.
+# 
+
+print("Initialize Model...")
+# Construct Model
+model = models.resnet18(pretrained=False).cuda()
+# Make model DistributedDataParallel
+model = torch.nn.parallel.DistributedDataParallel(model, device_ids=dp_device_ids, output_device=local_rank)
+
+# define loss function (criterion) and optimizer
+criterion = nn.CrossEntropyLoss().cuda()
+optimizer = torch.optim.SGD(model.parameters(), starting_lr, momentum=0.9, weight_decay=1e-4)
+
+
+######################################################################
+# Initialize Dataloaders
+# ~~~~~~~~~~~~~~~~~~~~~~
+# 
+# The last step in preparation for the training is to specify which
+# dataset to use. Here we use the `STL-10
+# dataset <https://cs.stanford.edu/~acoates/stl10/>`__ from
+# `torchvision.datasets.STL10 <https://pytorch.org/docs/stable/torchvision/datasets.html#torchvision.datasets.STL10>`__.
+# The STL10 dataset is a 10 class dataset of 96x96px color images. For use
+# with our model, we resize the images to 224x224px in the transform. One
+# distributed training specific item in this section is the use of the
+# ``DistributedSampler`` for the training set, which is designed to be
+# used in conjunction with ``DistributedDataParallel`` models. This object
+# handles the partitioning of the dataset across the distributed
+# environment so that not all models are training on the same subset of
+# data, which would be counterproductive. Finally, we create the
+# ``DataLoader``'s which are responsible for feeding the data to the
+# processes.
+# 
+# The STL-10 dataset will automatically download on the nodes if they are
+# not present. If you wish to use your own dataset you should download the
+# data, write your own dataset handler, and construct a dataloader for
+# your dataset here.
+# 
+
+print("Initialize Dataloaders...")
+# Define the transform for the data. Notice, we must resize to 224x224 with this dataset and model.
+transform = transforms.Compose(
+    [transforms.Resize(224),
+     transforms.ToTensor(),
+     transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])
+
+# Initialize Datasets. STL10 will automatically download if not present
+trainset = datasets.STL10(root='./data', split='train', download=True, transform=transform)
+valset = datasets.STL10(root='./data', split='test', download=True, transform=transform)
+
+# Create DistributedSampler to handle distributing the dataset across nodes when training
```

Comment:
Also add a comment saying this needs to be called after torch.init_process_group is called